services/shhext/mailservers/connmanager_test.go

```diff
@@ -0,0 +1,301 @@
+				output <- ev
+			}
+		}
+	})
+}
+
+func newFakeServer() *fakePeerEvents {
+	srv := newFakePeerAdderRemover()
+	srv.input = make(chan *p2p.PeerEvent, 20)
+	return srv
+}
+
+type fakeEnvelopeEvents struct {
+	input chan whisperv6.EnvelopeEvent
+}
+
+func (f fakeEnvelopeEvents) SubscribeEnvelopeEvents(output chan<- whisperv6.EnvelopeEvent) event.Subscription {
+	return event.NewSubscription(func(quit <-chan struct{}) error {
+		for {
+			select {
+			case <-quit:
+				return nil
+			case ev := <-f.input:
+				// will block the same way as in any feed
+				output <- ev
+			}
+		}
+	})
+}
+
+func newFakeEnvelopesEvents() fakeEnvelopeEvents {
+	return fakeEnvelopeEvents{
+		input: make(chan whisperv6.EnvelopeEvent),
+	}
+}
+
+func getNRandomNodes(t *testing.T, n int) map[enode.ID]*enode.Node {
+	rst := map[enode.ID]*enode.Node{}
+	for i := 0; i < n; i++ {
+		n, err := RandomeNode()
+		require.NoError(t, err)
+		rst[n.ID()] = n
+	}
+	return rst
+}
+
+func mergeOldIntoNew(old, new map[enode.ID]*enode.Node) {
+	for n := range old {
+		new[n] = old[n]
+	}
+}
+
+func TestReplaceNodes(t *testing.T) {
+	type testCase struct {
+		description string
+		old         map[enode.ID]*enode.Node
+		new         map[enode.ID]*enode.Node
+		target      int
+	}
+	for _, tc := range []testCase{
+		{
+			"InitialReplace",
+			getNRandomNodes(t, 0),
+			getNRandomNodes(t, 3),
+			2,
+		},
+		{
+			"FullReplace",
+			getNRandomNodes(t, 3),
+			getNRandomNodes(t, 3),
+			2,
+		},
+	} {
+		t.Run(tc.description, func(t *testing.T) {
+			peers := newFakePeerAdderRemover()
+			replaceNodes(peers, tc.target, peers.nodes, nil, tc.old)
+			require.Len(t, peers.nodes, len(tc.old))
+			for n := range peers.nodes {
+				require.Contains(t, tc.old, n)
+			}
+			replaceNodes(peers, tc.target, peers.nodes, tc.old, tc.new)
+			require.Len(t, peers.nodes, len(tc.new))
+			for n := range peers.nodes {
+				require.Contains(t, tc.new, n)
+			}
+		})
+	}
+}
+
+func TestPartialReplaceNodesBelowTarget(t *testing.T) {
+	peers := newFakePeerAdderRemover()
+	old := getNRandomNodes(t, 1)
+	new := getNRandomNodes(t, 2)
+	replaceNodes(peers, 2, peers.nodes, nil, old)
+	mergeOldIntoNew(old, new)
+	replaceNodes(peers, 2, peers.nodes, old, new)
+	require.Len(t, peers.nodes, len(new))
+}
+
+func TestPartialReplaceNodesAboveTarget(t *testing.T) {
+	peers := newFakePeerAdderRemover()
+	old := getNRandomNodes(t, 1)
+	new := getNRandomNodes(t, 2)
+	replaceNodes(peers, 1, peers.nodes, nil, old)
+	mergeOldIntoNew(old, new)
+	replaceNodes(peers, 1, peers.nodes, old, new)
+	require.Len(t, peers.nodes, 1)
+}
+
+func TestConnectionManagerAddDrop(t *testing.T) {
+	server := newFakeServer()
+	whisper := newFakeEnvelopesEvents()
+	target := 1
+	connmanager := NewConnectionManager(server, whisper, target)
+	connmanager.Start()
+	defer connmanager.Stop()
+	nodes := []*enode.Node{}
+	for _, n := range getNRandomNodes(t, 3) {
+		nodes = append(nodes, n)
+	}
+	connmanager.Notify(nodes)
+	var initial enode.ID
```

Comment:
I think having comments clearly separating the conceptual sections and providing a short explanation of what is being done would be a good addition, so that future developers can quickly glance over the code and understand the intent (i.e. `// Select and add 3 random nodes`, `// Wait 100ms and assert they were correctly added`, `// Notify that existing peer was dropped`, etc.)